sched/event/event_post.c

```diff
@@ -0,0 +1,142 @@
+
+#include <nuttx/event.h>
+#include <nuttx/sched.h>
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_post_internal
+ ****************************************************************************/
+
+static inline_function int nxevent_post_internal(FAR sem_t *sem)
+{
+  int semcount;
+
+  nxsem_get_value(sem, &semcount);
+  if (semcount < 1)
+    {
+      return nxsem_post(sem);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_post
+ *
+ * Description:
+ *   Post one or more events to an event object.
+ *
+ *   This routine posts one or more events to an event object. All tasks
+ *   waiting on the event object event whose waiting conditions become
+ *   met by this posting immediately unpend.
+ *
+ *   Posting differs from setting in that posted events are merged together
+ *   with the current set of events tracked by the event object.
+ *
+ * Input Parameters:
+ *   event  - Address of the event object
+ *   events - Set of events to post to event
+ *          - Set events to 0 will be considered as any,
+ *            waking up the waiting thread immediately.
+ *
+ * Returned Value:
+ *   This is an internal OS interface and should not be used by applications.
+ *   It follows the NuttX internal error return policy:  Zero (OK) is
+ *   returned on success.  A negated errno value is returned on failure.
+ *
+ * Assumptions:
+ *   This function may be called from an interrupt handler.
+ *
+ ****************************************************************************/
+
+int nxevent_post(FAR nxevent_t *event, nxevent_mask_t events)
+{
+  FAR struct tcb_s *stcb;
+  irqstate_t flags;
+  int ret = 0;
+
+  if (event == NULL)
+    {
+      return -EINVAL;
+    }
+
+  flags = enter_critical_section();
+
+  /* NOTE: Current implementation does not support multiple waiters. */
+
+  stcb = (FAR struct tcb_s *)dq_peek(SEM_WAITLIST(&event->sem));
+  if (stcb != NULL)
+    {
+      if (events == 0)
+        {
+          events = stcb->events ? stcb->events : ~0;
+        }
+
+      /* Specified events */
+
+      if (stcb->events)
+        {
+          events &= stcb->events;
+          event->events |= events;
+          if ((stcb->events & event->events) == stcb->events)
+            {
+              ret = nxevent_post_internal(&event->sem);
+            }
+        }
+
+      /* Any events */
+
+      else
+        {
+          event->events |= events;
+          ret = nxevent_post_internal(&event->sem);
+        }
+    }
+  else
+    {
+      if (events == 0)
```

Comment:
should we remove the special handling of events==0? it isn't not good to set all flags sometime, but some bit other time.

Reply:
still not resolve yet. it's better to remove this special path